packages/flutter_test/lib/src/finders.dart

```diff
@@ -460,57 +469,544 @@ class CommonFinders {
+    } else {
+      if (!hasFound) {
+        find();
+      }
+      return found.toString();
+    }
+  }
+}
+
+/// The results of searching with a [FinderBase].
+class FinderResult<TCandidate> extends Iterable<TCandidate> {
+  /// Creates a new [FinderResult] that describes the [values] using the given
+  /// [requirements].
+  ///
+  /// This description should be a brief English noun phrase describing the
+  /// finder's requirements. It should make sense as the end of the sentence
+  /// "Found an item with <description>".
+  ///
+  /// As an example, a `FinderBase<Object>` that searches for a candidate of a
+  /// specific type would provide 'type MyObject', so that the full sentence
+  /// ends up as "Found an item with type MyObject."
+  FinderResult(String requirements, Iterable<TCandidate> values)
+  : _requirements = requirements, _values = values;
+
+  final String _requirements;
+  final Iterable<TCandidate> _values;
+
+  @override
+  Iterator<TCandidate> get iterator => _values.iterator;
+
+  @override
+  String toString() {
+    final List<TCandidate> valuesList = _values.toList();
+    // This will put each value on its own line with a comma and indentation
+    final String valuesString = valuesList.fold(
+      '',
+      (String current, TCandidate candidate) => '$current\n  $candidate,',
+    );
+    return 'Found ${valuesList.length} ${TCandidate}s $_requirements: ['
+    '$valuesString'
+    '\n]';
+  }
+}
+
+/// A mixin that provides a standard implementation of [FinderBase.allCandidates]
+/// for [Element] [FinderBase]s.
+mixin ElementFinderMixin on FinderBase<Element> {
+  /// Whether this finder includes nodes that are offstage.
+  ///
+  /// If this is false, then the elements are walked using
   /// [Element.debugVisitOnstageChildren]. This skips offstage children of
   /// [Offstage] widgets, as well as children of inactive [Route]s.
-  final bool skipOffstage;
+  bool get includeOffstage;
 
   /// Returns all the [Element]s that will be considered by this finder.
   ///
   /// This is the internal API for the [Finder]. To obtain the elements from
   /// a [Finder] in a test, consider [WidgetTester.elementList].
   ///
   /// See [collectAllElementsFrom].
-  @protected
+  @override
   Iterable<Element> get allCandidates {
     return collectAllElementsFrom(
       WidgetsBinding.instance.rootElement!,
-      skipOffstage: skipOffstage,
+      includeOffstage: includeOffstage,
+    );
+  }
+
+  /// Returns a variant of this finder that only matches elements reachable by
+  /// a hit test.
+  ///
+  /// The [at] parameter specifies the location relative to the size of the
+  /// target element where the hit test is performed.
+  FinderBase<Element> hitTestable({ Alignment at = Alignment.center }) => _HitTestableFinder(this, at);
+}
+
+/// A base class for creating [Element] [FinderBase]s.
+abstract class ElementFinder extends FinderBase<Element> with ElementFinderMixin {
```

Comment:
Why is there the ElementFinderMixin and this abstract class? Wouldn't just having the abstract `ElementFinder` class be enough? What does the mixin enable?

Reply:
I removed the mixins for `FinderBase<Element>` and `FinderBase<SemanticsNode>`, but left the ones for Chained, Match, and a few internals where the mixin logic could be reused across both semantics and widget finders.